test/test_dispatch.py

```diff
@@ -0,0 +1,359 @@
+                active_ops.remove(op_ix)
+                # NB: this finally test asserts that if a registrations fails,
+                # the dispatcher is left in the same state *that it was before*!
+                check_invariants(
+                    "running ctors {} and then failing to run ctor {} "
+                    "(did this failure leave the dispatcher in a wedged state? "
+                    "it shouldn't!)"
+                    .format(ctor_order[:i], op_ix))
+                break
+        last_ctor = i
+        if raises and len(active_ops) == len(ops):
+            # Destroy references first, as some test frameworks (like pytest)
+            # will retain references in the exception raised by assertTrue! EW!
+            refs = None
+            self.assertTrue(
+                False,
+                "expected exception to be raised, but nothing was raised "
+                "(after running ctors {})".format(ctor_order))
+        # In the order specified by dtor_order, run deregistrations
+        for i, op_ix in enumerate(dtor_order):
+            # Trigger a destruction
+            refs[op_ix] = None
+            # discard not remove, since we may not have actually deregistered
+            # anything if there was an error raised
+            if raises:
+                active_ops.discard(op_ix)
+            else:
+                active_ops.remove(op_ix)
+            check_invariants(
+                "running ctors {}, then running dtors {}"
+                .format(ctor_order[:last_ctor + 1], dtor_order[:i + 1])
+            )
+        return results[set_to_report][0]
+
+    # Operator registrations are commutative (as static initializers can
+    # run in any order) and invertible (by deregistration).  (Subject
+    # to some caveats: some legacy behavior in the system are not commutative--
+    # we want to get rid of these!)
+    #
+    # So while in principle we could simply test a set of operations
+    # by just running them one by one in the order specified by the user,
+    # we can get more assurance about these extra properties by doing
+    # more work:
+    #
+    # 1. Don't run the registrations once in a fixed order: run every possible
+    #    permutation.  Similarly, run every permutation of deregistration order.
+    #
+    # 2. Don't just check the end state of the dispatcher: for every
+    #    subset of operator registrations, ensure that the computed
+    #    intermediate state is path independent.  One thing to note:
+    #    in this function, we assume each operation is unique.  In general,
+    #    there may be duplicated registrations, but these are usually
+    #    idempotent or legacy.  We test for behavior here separately.
+    #
+    # NB: checking all permutations means this function is exponential in
+    # the length of ops!  So don't pass too many ops to this function!
+    def commute(self, name, ops, ctor_order=None, raises=False):
```

Comment:
Do you think it makes sense to call the registered ops from python and verify their outputs to make sure everything is wired up correctly?

Reply:
I'm relying on dumpState to tell me if the functions are registered correctly. There's some calling tests in C++, but I couldn't figure out a way to conveniently make invocations from Python and setup some sort of mocking API, so I decided not to do it.